src/transformers/models/owlvit/modeling_owlvit.py

```diff
@@ -0,0 +1,1418 @@
+        image_embeds: torch.FloatTensor,
+        query_embeds: torch.FloatTensor,
+        query_mask: torch.Tensor,
+    ) -> Tuple[torch.FloatTensor]:
+
+        image_class_embeds = self.dense0(image_embeds)
+
+        # Normalize image and text features
+        image_class_embeds /= torch.linalg.norm(image_class_embeds, dim=-1, keepdim=True) + 1e-6
+        query_embeds /= torch.linalg.norm(query_embeds, dim=-1, keepdim=True) + 1e-6
+
+        # Get class predictions
+        pred_logits = torch.einsum("...pd,...qd->...pq", image_class_embeds, query_embeds)
+
+        # Apply a learnable shift and scale to logits
+        logit_shift = self.logit_shift(image_embeds)
+        logit_scale = self.logit_scale(image_embeds)
+        logit_scale = self.elu(logit_scale) + 1
+        pred_logits = (pred_logits + logit_shift) * logit_scale
+
+        if query_mask is not None:
+            if query_mask.ndim > 1:
+                query_mask = torch.unsqueeze(query_mask, dim=-2)
+
+            pred_logits = pred_logits.to(torch.float64)
+            pred_logits = torch.where(query_mask == 0, -1e6, pred_logits)
+            pred_logits = pred_logits.to(torch.float32)
+
+        return (pred_logits, image_class_embeds)
+
+
+class OwlViTImageTextEmbedder(nn.Module):
+    def __init__(self, config: OwlViTConfig):
+        super().__init__()
+
+        self.clip = OwlViTModel(config)
+        self.layer_norm = nn.LayerNorm(config.vision_config.hidden_size)
+
+    def forward(
+        self,
+        pixel_values: Optional[torch.FloatTensor] = None,
+        input_ids: Optional[torch.Tensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        output_attentions: Optional[bool] = None,
+    ) -> Tuple[torch.FloatTensor, torch.FloatTensor]:
+
+        image_embeds, text_embeds = None, None
+
+        # Encode text
+        if input_ids is not None:
+            text_embeds = self.clip.get_text_features(
+                input_ids=input_ids, attention_mask=attention_mask, output_attentions=output_attentions
+            )
+
+        # Encode image
+        if pixel_values is not None:
+            image_embeds = self.clip.get_image_features(
+                pixel_values, return_projected=False, output_attentions=output_attentions
+            )
+
+            # Resize class token
+            new_size = tuple(np.array(image_embeds.shape) - np.array((0, 1, 0)))
+            class_token_out = torch.broadcast_to(image_embeds[:, :1, :], new_size)
+
+            # Merge image embedding with class tokens
+            image_embeds = image_embeds[:, 1:, :] * class_token_out
+            image_embeds = self.layer_norm(image_embeds)
+
+        return (image_embeds, text_embeds)
+
+
+class OwlViTForObjectDetection(OwlViTPreTrainedModel):
+    config_class = OwlViTConfig
+    main_input_name = "pixel_values"
+
+    def __init__(self, config: OwlViTConfig):
+        super().__init__(config)
+
+        self.embedder = OwlViTImageTextEmbedder(config)
```

Comment:
```suggestion
        self.owlvit = OwlViTModel(config)
```
Something we have in every model in the Transformers library, is that the head model directly uses the non-head model as an attribute, for instance `BertForSequenceClassification` uses `self.bert = BertModel(config)`, another example is `DetrForObjectDetection` which define `self.detr = DetrModel(config)`. The model then adds the head on top.

This allows to load a head model with the weights of a pre-trained base model, thanks to the `base_model_prefix`. However, I currently see that the `base_model_prefix` is set to "owlvit", but that means it won't work as intended.

Suppose you've trained a CLIP model yourself (an OwlViTModel), and then you want to fine-tune an OwlViTForObjectDetection model. Ideally, one should be able to do:

model = OwlViTForObjectDetection.from_pretrained("nielsr/my-owl-vit-pretrained-model")

This will then instantiate the weights of the "base", and randomly initialize the weights of the head.

Reply:
I see, to expand on that:`OwlViTModel` is an alias of CLIPModel and is trained separately from scratch in the first stage and then the whole model (base + embedder + object detection model) is fine-tuned.

Right now, `OwlViTModel` is an attribute of the `OwlViTImageTextEmbedder` which merges class tokens with image embedding outputs of CLIP and applies layer normalization. I can merge `OwlViTImageTextEmbedder` with `OwlViTForObjectDetection` but it is a big change and the authors haven't published the training code yet.

@sgugger approved the PR, is it ok if I wait until implementing the training code for this change?

Reply:
We can't merge it and change afterwards, as it will not be backwards compatible. 

I've looked at the code and I'm not really seeing the benefit of `self.embedder`. It's not used in the forward pass either. Would it be possible to just have self.clip and self.layernorm inside the `ForObjectDetection` model?

I'm also wondering why `self.image_embedder` and `self.text_embedder` are called separately.

Reply:
Yes, I'm working on moving self.clip and self.layernorm inside the `ForObjectDetection` model.

As for calling self.image_embedder and self.text_embedder separately, I was conforming to the original repo but I agree that it'd make more sense to make a single call. I'll update it accordingly